soroban-sdk-macros/src/derive_struct.rs

```diff
@@ -25,7 +25,11 @@ pub fn derive_type_struct(
 ) -> TokenStream2 {
     // Collect errors as they are encountered and emit them at the end.
     let mut errors = Vec::<Error>::new();
-
+    let allow_hash = if let Visibility::Public(_) = vis {
+        false
+    } else {
+        true
+    };
```

Comment:
If I understand correctly these lines, and the similar lines in the other macros, means folks can use `Hash` in `contracttype`s that are not exported. I think the intent is to allow `Hash` on internal only types.

Unfortunately you can still use non-pub contracttype's on public functions. So it'd still be possible to write code that accepts a Bytes, which becomes a Hash, and is now unsafe.

I think we can start out with limited support for Hash, and make it so the only places you can use it are explicitly `__check_auth` and the output of the hash functions. 

Ideally we'd have a way to allow people to store the hash, but I don't see a way to do that reliably without the Soroban Env supporting Hash as a first-class type.

Reply:
You're right. I did not realize the non-pub types can be used in public function. In that case, the visibility check -> allow_hash doesn't make sense. 

On the other hand,  (not part of this work) we should probably add some visibility restrictions to contract types, otherwise it is quite error-prone since the users might be expecting some safety guarantees from non-pub types.